### ISSUE TITLE
Apply functions from SparkR

### ISSUE THREAD

**Issue by ElianoMarques** [rstudio/sparklyr#81]
How would we do something similar to the apply functions in SparkR?

``` r
dapply(x, func, schema)
spark.lapply(sc, list, func)
```

[Documentation can be found here](http://spark.apache.org/docs/2.0.0-preview/api/R/index.html).

Thank you

**Comment by javierluraschi** [comment]
@ElianoMarques arbitrary r-code is currently not supported in `sparklyr`. However, we do have plans to bring this in the near future. I'll keep this issue open to provide updates.

**Comment by ElianoMarques** [comment]
Thanks for the reply.

**Comment by ElianoMarques** [comment]
@javierluraschi So for example, lets say you're interested in performing a particular ML model in a  
 partition or a particular node from R but that particular algorithm is not available in the ML framework of both ml and h20? The solution could require classic ML + Ensembling. 
There are out there a few ML processes that doesn't have a 1-2-1 with ML or H20 but are very valuable. TensortFlow (great project you guys started), SOMs, Time Series. Other areas this could help are simulation, automation of model selection and the list would continue to go on.

**Comment by kykuo** [comment]
Wanted to throw out possibly incorporating `purrr` functions... will have to think about how everything plays together.

**Comment by ElianoMarques** [comment]
Thanks @kykuo, that would be great.

**Comment by javierluraschi** [comment]
Received mail with another request to add support for parallel execution through dplyrs `do`: https://cran.r-project.org/web/packages/dplyr/dplyr.pdf - "Do arbitrary operations on a tbl." section. Noted! Enabling this feature would open up a bunch of interesting scenarios.

**Comment by hafen** [comment]
+1000 for arbitrary r-code execution through `do`.

**Comment by AjarKeen** [comment]
`group_by() %>% do()` would enable a wide range of possibilities. The SparkR UDF implementations have constraints that prevent them from meeting my needs.

**Comment by javierluraschi** [comment]
@AjarKeen could you share some examples of current limitations with SparkR UDF to make sure this gets implemented without them?

**Comment by AjarKeen** [comment]
Sure! The SparkR UDF implementation comes with some pretty hard constraints. 

1. SparkR UDFs are only allowed to receive 2-3 arguments: a key, an R data frame, and (in some cases) a schema for the output. Your function definition can have other arguments if they have defaults, but you can't pass those arguments yourself at runtime. 

2. These UDFs are only allowed to return a data frame. This makes it very hard to implement e.g. custom model training methods where the thing you want to return for each group is a model object, not a data frame. (I tried flattening my objects into DFs to reconstruct them later, but that didn't work well at all.)

3. The environment of the UDF has no access to the R environment from which the SparkR *apply function was called. The only way I've been able to get access to package functions inside my SparkR UDF is to `::` them. (Some things from the originating R environment do get passed along: sourcing .R files that contain functions does make those functions available inside the UDF. But I've also tried defining variables in my originating R environment and asking for them in my UDF to get around item 4, and that doesn't work.)

4. Side effects are difficult to achieve. The R instance is invoked by another user (e.g. on AWS in YARN mode, `system("whoami")` identifies me as 'yarn', which means no privileges to write to the filesystem outside of the Hadoop scratch directory, no privileges to copy objects to S3, etc, etc. 

5. Some things seem to get lost in translation. I implemented `...` in my model training function to pass a formula which then gets parsed by the actual learner, which works in a non-Spark local setting. When I put this inside a SparkR UDF, however, the character vector doesn't get passed down to the learner correctly. I can't be more specific than this because I haven't been able to identify the cause of the problem. :(

The actual thing I want to do is group a Spark DataFrame by a key, train a custom R model (comprised of multiple custom submodels) for each group, and then either get the model object back for each group, or save it somewhere. I can do that now in a limited way, but once my model becomes at all complex, I start hitting the limitations I described above and it stops working.

**Comment by cosmincatalin** [comment]
Just an opinion: due to the distributed nature of Spark, I find it really challenging, but really awesome, but really challenging, but really awesome if the `do` function gets implemented at least on par with the SparkR `*apply` functions.

Issue number _1_ above is a valid feature request, but can be worked around by modifying the data frame argument in advance to contain runtime information. Not pretty, but it would work.

Issue number _2_ means a total restructure of the way _sparklyr_ works, which is based on data frames, and what one would expect.

Issue number _3_ touches a little bit on how Spark works, depending on the type of objects you have in your parent scope, using them in the closure might be outright impossible because of serialization issues. Just being able to use _some_ outer scope variables is pretty cool in my world.

Issue number _4_ looks like is unrelated to _SparkR_.

**Comment by hadley** [comment]
FWIW, I no longer consider `do()` to be the way forward for arbitrary code execution. I think list-cols + `mutate` + a multidplyr approach (or parallel purrr) to be a more likely way forward.

**Comment by AjarKeen** [comment]
I'm all for best practices. I'm also willing to contribute to this effort if there's anything useful I can do, because if we can get UDFs to work in a more flexible/robust and R-like way than SparkR does, it will be tremendously useful for my job. For example, I'd be happy to run distributed tests on AWS for new sparklyr features such as this one.

**Comment by akzaidi** [comment]
Happy to help with Azure on HDI tests as well :)

Just my two cents: while the list-colsts and `mutate()` approach might be more pleasing from a tidy perspective, I think it'll generally be hard to have nested DataFrames in Spark, without at least braking the API and resorting to RDDs. I think a `do()` or `purrr:map()` approach would be a lot easier and would still maintain the proper functional aspects of developing UDFs.

**Comment by AjarKeen** [comment]
I think `map` is a very smart approach, since it would mirror the Scala and Python APIs.

**Comment by hafen** [comment]
Here are some thoughts / ideas supporting @hadley's comments (I think!).

I think the approach that would be the easiest to implement and hit the most use cases would be if we could stick with Spark DataFrames. But in my opinion, any kind of UDF support automatically means support for list-columns / arbitrary R objects stored in columns of a Spark DataFrame. However, I think there is an easily-with-reach approach to this provided the following assumptions regarding nested data frames / list-columns:

1. There is never any need to do certain SQL-like operations (e.g. via filter, arrange) on the nested data frames or list-colum objects (operations that only make sense for atomic data structures).
2. Pretty much all list-column objects we create are probably going to be things that only R understands (e.g. model fits like lm objects, etc.). Otherwise why would we be creating them? Therefore we don't care if other languages can't handle these columns because even if they could they wouldn't know what to do with them.
3. Related to (2), the use of list-columns / nested data frames will probably typically be for intermediate data structures that we manipulate in various ways with the ultimate goal to get back to a dataset or result that can be stored back into a cross-platform friendly DataFrame or read back in to R
4. Given 1-3, users should be entirely free to specify any kind of R object for a list-column and not have to care about specifying its schema (requiring anything like this would be a nightmare and people wouldn't use it)

Given these assumptions, and the case to stick with Spark DataFrames, I would propose the following for any specified mutation (thinking of both nest->mutate and group_by->summarise):

- If a mutation uses only data and operations that can be understood by SparkSQL, handle it as usual
- If a mutation has a UDF (basically anything outside supported SparkSQL operations), pipe the input colums it requires into an R RPC for each group and process it
  - If the result of this mutation is a standard SparkSQL data type, treat it as such
  - If the result is an R nested data frame or list-column, store a serialized version (binary/string type? - hopefully can be done efficiently!) and accept that this column is now relegated to subsequent R UDF computation

There are details of the RPC such as how all the data is gathered together for each group that will have some challenges. The way we are used to thinking of "group_by + custom mutate" is that our mutation function expects to operate on all the data for a group. We can't guarantee that all the data for a group will be of manageable size with larger distributed data. And it's too much (and often impossible) to ask the user to specify a mutate function that can be iteratively updated with new data. So I wonder if some kind of analyze tool can be provided to help users understand if their grouping + UDF will be feasible or not. On the flipside there can also be the case where groups are tiny and the number of groups insanely huge, where you would want to take a multidplyr approach mentioned by Hadley and send chunks of groups into an RPC.

Given functionality for list-columns, UDF support, and group_by, I think the vast majority of distributed computations and data structures can be provided for, continuing under the dplyr / tidy paradigm that people are familiar with and without having to resort to RDDs, MapReduce-like stuff, etc., although I may be naively missing something.

I used to think sticking to data frames and SQL-like operations was severely limiting, but with UDF support and list-columns, I've come around (see [here](https://speakerdeck.com/hafen/the-need-for-flexibility-in-distributed-computing-with-r)).

**Comment by hadley** [comment]
@hafen that aligns very closely with my thinking :)

**Comment by codingbutstillalive** [comment]
Hi!

Sparklyr is a wonderful approach to Spark. I have to admit that I do not understand everything that was discussed here, but I do want to point out that even in R, not every interesting data item is a data frame. For example, I often have to do text processing, where I need to handle large collections of text documents, and I need to split those into tokens and things like this. I think that this is a strength of the scala and python interfaces for Spark, which allow to operate on text files in much the same ways as on cvs files. That is what I can see from the map().reduce() commands in scala, like for example in "val parsed = noheader.map(line => parse(line))".

From the O'Reilly book, Advanced analytics in Spark, I understood that constructs like:

"val merged = nas.reduce((n1, n2) => {  n1.zip(n2).map { case (a, b) => a.merge(b) } })" 

are extremely useful and, most of all, efficient to use. Will we exclude such flexibility in R with the discussed design decisions, if Sparklyr becomes a standard? I do not understand enough from the details, to assess it, honestly.

I understood that sparklyr wants to implement the concept of tidy data frames; however, on the other hand you are also aiming to become THE spark interface for R, that is my impression. I wonder if this can work if you neglect other data structures than tabular ones. 

I might be wrong with that impression. Please correct me, if I am. I do not want to disturb the discussions, just give a perspective from a newbie on the topic.

Best regards, Sebastian

**Comment by javierluraschi** [comment]
Here is a reference that Sebastian sent by mail: https://forums.databricks.com/questions/10251/how-to-count-number-of-missing-values-for-each-col.html

**Comment by hadley** [comment]
I think we'd need both interfaces for tidy data frames and lists/vectors.  (But see http://tidytextmining.com for handling text data with data frames)

**Comment by ElianoMarques** [comment]
@javierluraschi any expectation on when we could have this in sparklyr? We are about to start a piece of work and might have to adopt sparkR given this feature its there. it would be a shame given the investment made in getting up to speed with sparklyr across the team?

**Comment by zero323** [comment]
@cosmincatalin

> Issue number 4 looks like is unrelated to SparkR.

To some extent it is highly related to Spark execution model. Any kind of side effects can be extremely hard to manage in complex pipelines. What happens when you enable speculative executing, how to handle task failures, how to deal with normal task re-execution? If you wan to allow for side effects there is a lot of low level thinking required to do it right. 

That being said, AFAIK `local-user` on Yarn is configurable. 

@AjarKeen 

> SparkR UDFs are only allowed to receive 2-3 arguments: a key, an R data frame, and (in some cases) a schema for the output. Your function definition can have other arguments if they have defaults, but you can't pass those arguments yourself at runtime.

You can easily use currying / partial application for that so it is hardly a limitation.

```R
schema <-   structType(
    structField("species", "string"),
    structField("max_s_width", "double"))

f <-  function(i)  function(k, x) {
  dplyr::summarize(dplyr::group_by(x, Species), max(Sepal_Width * i))
}

as.DataFrame(iris) %>% gapply("Species", f(3), schema) %>% head
```

> Some things seem to get lost in translation. I implemented `...` in my model training function to pass in a formula which then gets parsed by the actual learner, which works in a non-Spark local setting. When I put this inside a SparkR UDF, however, the character vector doesn't get passed down to the learner correctly

This should probably deserves a JIRA ticket but in general determining a minimal set of objects which have to be serialized as a part of the closure can be tricky. The more opaque is your code, the harder it gets.

**Comment by javierluraschi** [comment]
@ElianoMarques no updates on this yet, but I do understand the urgency of this request. Thanks for the kind reminder to get this going, I will update this issue when we get this started.

**Comment by codingbutstillalive** [comment]
That's great! I am also still hoping for it. Would be a really great addition.

**Comment by codingbutstillalive** [comment]
Want to keep this alive. It's still urgently needed.

**Comment by codingbutstillalive** [comment]
I wonder. Are you guys working on this, or did you give it up?

**Comment by javierluraschi** [comment]
@codingbutstillalive we will keep it alive, I have a prototype repo to investigate this, shouldn't be that hard to get basic parallelization working: https://github.com/javierluraschi/sparkworker - My plan it to get this working there and then merge back to `sparklyr`, not usable at this point in time, ping me back in a couple weeks.

**Comment by codingbutstillalive** [comment]
That's wonderful news. I appreciate that. Good luck and success!

**Comment by ElianoMarques** [comment]
Awesome news @javierluraschi . Thanks for that.

**Comment by codingbutstillalive** [comment]
I hope you are still motivated ;)

**Comment by codingbutstillalive** [comment]
Seems to be more difficult than expected, I suppose. Right?

**Comment by kevinushey** [comment]
We're still working on this; I think we'll be able to accomplish something in the near future but given that we're balancing a large number of responsibilities it's possible this will still take a little longer than expected.

**Comment by codingbutstillalive** [comment]
I hope for the best.

**Comment by codingbutstillalive** [comment]
Just to let you know. Once more, I was asked to switch from SparkR to sparklyr for our data science courses, but needed to refuse it because of those insufficiencies. It's a pity.

**Comment by chezou** [comment]
Hey @codingbutstillalive don't be so noisy. sparklyr is an OSS. If you need some functions, why not contribute to them? 

For example, there is an implementation of the concept.
https://github.com/javierluraschi/sparkworker

It's a plugin of sparklyr, so why don't you try it and report any problem?

I'm not a commiter of sparklyr, but it would be nice for us to contribute the project 😉

**Comment by codingbutstillalive** [comment]
@chezou I  consider the requested functionality as mandatory. Plugins should be reserved for convenience or special function imho.

**Comment by codingbutstillalive** [comment]
@chezou By the way, it's about the sparkworker plugin that we were talking about. It was supposed to address the problem of missing apply functions. See comments from 14th of March above.

**Comment by francoisjehl** [comment]
Hi @javierluraschi , I've been playing a bit with your extension. I was wondering if you thought about using a managed JVM based engine such as Renjin for R code execution, instead of relying on forking RScript. In my company we run Spark mostly within Yarn, and try to keep dependencies on the physical workers as low as possible (in other words: we'll never install R on our 1000+ Hadoop data node :) )

**Comment by chezou** [comment]
I didn't try, but I guess we can use this kind of trick: https://www.continuum.io/blog/developer-blog/conda-spark

Basic idea is distributing R binary with conda environment and set `spark.r.command`.
https://github.com/javierluraschi/sparkworker/blob/master/java/process.scala#L28

I guess this idea may help your situation.

**Comment by codingbutstillalive** [comment]
I see from javiers recent commits to sparkworker and the activities here that the topic has become revived. That is fantastic, guys, thanks for your efforts.

**Comment by hafen** [comment]
I don't recall if I have already mentioned this or not, and maybe you are already doing something similar, but one approach we created and highly advocate with the use of [RHIPE](https://github.com/delta-rho/RHIPE) is to create an archive of your local R environment and all its dependencies and ship it to the nodes using Hadoop's distributed cache (I would assume Spark's broadcast variables could be used similarly).

With this approach you don't need to worry about having R and the correct packages installed on all the nodes. Further, you only need to create and broadcast the archive whenever you have significantly updated your R environment. Otherwise, it's cached and available for subsequent jobs.

Here's the source for this functionality: https://github.com/delta-rho/RHIPE/blob/master/src/main/R/bashRhipeArchive.R. The downsides are that it requires bash and assumes the same architecture on all nodes.

**Comment by javierluraschi** [comment]
@francoisjehl Thanks for the comment. What would it take to convince you that it's worth installing R in each machine? Python is probably installed, isn't it? `Renjin` looks great as an engine, but the challenge would be to bring all the packages to run into it; I'm sure you are aware that popular CRAN packages use native code, etc. so those packages would be hard to port to `Renjin`. That said, I wouldn't be opposed supporting `Renjin` at some point for those clusters that are locked from installing additional binaries, I'll keep this in mind.

@hafen Hi, I don't think you've mentioned it. But I agree, we might end up having to do something like `RHIPE`: https://github.com/delta-rho/RHIPE/blob/master/src/main/R/bashRhipeArchive.R#L49-L55 - So far, I'm assuming each node has the right packages, once that works, we can tackle this case which I'm sure will be pretty common.

**Comment by codingbutstillalive** [comment]
From my practical experience, it is possible (but difficult) to convince the cluster admin to install R on each cluster node, but definitely not all the packages that are needed. This could become a bottleneck then. 

I think Javier has a good point. The RHIPE approach is a very good way to go once the general operations are running. I didn't imagine the whole thing to have so much dependencies. Now I understand how difficult it is. Great job, guys.

**Comment by hafen** [comment]
Sounds like a good approach. Yes the main driver for the shipping to the nodes is not necessarily the base R installation but the constant updating of packages that no cluster admin will ever be able to keep up with (not to mention different users with different environments). It turns out to be much hairier than just copying packages because of a lot of possible shared library dependencies, etc. (hence the rest of the fun stuff in the code I referenced).

**Comment by javierluraschi** [comment]
Here is the initial [Implement R Workers](https://github.com/rstudio/sparklyr/pull/728) PR to support a distributed apply function.

Notice that, currently, this requires your cluster to have the right packages installed or manually install them under the closure. I've opened https://github.com/rstudio/sparklyr/issues/729 to track and implement something similar as to what @hafen described.

I think it would be healthy to close this issue and open new suggestions, say support for `purrr::map`, `dplyr` integration, etc. in separate issues.

Thanks for all the feedback and sticking along while this got implemented.
